### PR TITLE
DSY-1994 Button update

### PR DESCRIPTION
**Pull request by lauralucca**
# Description

This deals with technical and functional debts of the **Button** component, including new features and visual updates, according to [DSY-1994 Update | Button [RN]](https://natura.atlassian.net/browse/DSY-1994), and listed below:
- [x] New Icon attribute
- [x] New Ripple animation
- [x] New Display styling documentation
- [x] Sizing and alignment fixes
- [x] Coding legibility and good practices improvements
- [x] Documentation changes for better illustration
- [x] Renaming of the Size attribute options (BREAKING CHANGE)
- [x] Removing of directory `android/app/src/main/assets/` from [.gitignore](https://github.com/natura-cosmeticos/natds-rn/pull/159/files#diff-bc37d034bad564583790a46f19d807abfe519c5671395fd494d8cce506c42947)
- [x] Removing of `IColors` interface from [src/common/themeSelectors/colors/colors.ts](https://github.com/natura-cosmeticos/natds-rn/pull/159/files#diff-6173c32bf3d57161386b99874a1f41369a8a511f137a86991e2071a2e3824e7d) and using directly the interface from `@naturacosmeticos/natds-themes`

No dependencies were added or removed for this change.

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

# How Has This Been Tested?

To see the visual changes included in this, you may want to run the sample phone app or the storybook web app and check under the `Button` component section.

- [x] All button examples should be styled as the UI documentation shows
- [x] All button examples should presente the ripple effect animation when clicked
- [x] There should be a section called **"variants"** showing examples for `contained`, `outlined` and `text` variants
- [x] There should be a section called **"icon"** showing examples of buttons in all variants with icons on the `left` and on the `right` of its label
- [x]  There should be a section called **"size"** showing examples of buttons in all variants and in `semi`, `semix` and `medium` sizes
- [x]  There should be a section called **"disabled"** showing examples of disabled buttons in all variants
- [x]  There should be a section called **"display"** showing examples of buttons in all variants showing the `display block` and `display inline` behaviours
- [x]  There should be a section called **"interactive"** where the user can pick and customize all button expected behaviours and attributes

To ensure the correct the behaviour and provide reliability the following automated tests were included and/or updated:

In [src/components/Button/Button.test.tsx](https://github.com/natura-cosmeticos/natds-rn/pull/159/files#diff-68b91b497740b3078f98243bb823ddc8a0264fe0178b965e04c081e98c58d7b3)
- [x] `should render the icon to the right by default'`
- [x] `should render the icon to the left`

In [src/components/TouchableRipple/TouchableRipple.test.tsx](https://github.com/natura-cosmeticos/natds-rn/pull/159/files#diff-6614fbf2f4ec8951c3cf622ea9a1e27b159d2d6b62b9fa96003dce41ca22a981)
- [x] `Should calculate the ripple size based on the biggest side of the element`

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: src/common/themeSelectors/sizes/sizes.ts [hunk 1]
```diff
@@ -21,30 +18,18 @@ export const getSize = (theme: Theme, size: keyof Size) => checkTheme(theme).siz
...
-      paddingTop: small,
+      paddingHorizontal: getSpacingSmall(theme),
     },
     medium: {
       minHeight: getSize(theme, 'semiX'),
-      paddingBottom: 12,
-      paddingLeft: small,
-      paddingRight: small,
-      paddingTop: 12,
+      paddingHorizontal: getSpacingSmall(theme) - getSpacingMicro(theme),
```
Comment (arielwb) [inline src/common/themeSelectors/sizes/sizes.ts:28]: a gente deve evitar esse tipo de calculo pq isso dificulta o controle dos tokens via tema.
Isso me parece um smell, ou nosso layout ta usando espaçamentos que nao temos tokens, ou a gente esta tentando abstrair algo que não ficou facil de entender.

File: src/components/Button/Button.test.tsx [hunk 2]
```diff
@@ -1,139 +1,107 @@
...
-import renderer from 'react-test-renderer';
-import { fireEvent, render } from '@testing-library/react-native';
+import { fireEvent } from '@testing-library/react-native';
 import { ThemeProvider } from 'styled-components/native';
 import theme from '../../common/themeSelectors/theme/mock-theme.json';
 import { Button, ButtonProps } from './Button';
+import { renderWithTheme } from '../../../test/testHelpers';
 
 jest.mock('react-native/Libraries/Components/Touchable/TouchableHighlight',
   () => 'TouchableHighlight');
```
Comment (arielwb) [inline src/components/Button/Button.test.tsx:10]: se deletarmos os mocks, estamos testando também o codigo dos seletores. Os mocks devem ser mantidos.
Alem disso, senti falta do mock do touchable ripple, pra nao testar ele também

File: src/components/Button/Button.test.tsx [hunk 3]
```diff
@@ -1,139 +1,107 @@
...
-    <Button {...props} />
-  </ThemeProvider>,
-));
-
 const defaultProps = ({
   onPress: () => {},
   text: 'label button',
 });
 
+const renderButton = (
```
Comment (arielwb) [inline src/components/Button/Button.test.tsx:16]: usar o `renderWithTheme`

File: src/components/Button/Button.tsx [hunk 4]
```diff
@@ -21,10 +22,18 @@ export interface ButtonProps {
...
    */
   disabled?: boolean
+  /**
+   * The icon name
+   */
+  iconName?: string,
+  /**
+   * Icon position relative to text label `left` | `right`
+   */
+  iconPosition?: IconPositions
```
Comment (arielwb) [inline src/components/Button/Button.tsx:32]: faltou testar as novas props

File: src/components/Button/ButtonBase.tsx [hunk 5]
```diff
@@ -1,43 +1,44 @@
-/* eslint-disable max-lines */
-import React from 'react';
+/* eslint-disable max-params, max-lines */
+import React, { useState } from 'react';
+import { Text, View } from 'react-native';
 import styled, { withTheme } from 'styled-components/native';
-import { NativeSyntheticEvent, NativeTouchEvent } from 'react-native';
 import {
   IColors,
```
Comment (arielwb) [inline src/components/Button/ButtonBase.tsx:6]: esse IColors tem que morrer!!!!!!!
Devemos usar as interfaces do natds-themes!!!!!

File: src/components/Button/ButtonBase.tsx [hunk 6]
```diff
@@ -1,43 +1,44 @@
...
   getShadowBySize,
   getColorByName,
 } from '../../common/themeSelectors';
+import { Icon, IconColors } from '../Icon';
+import { TouchableRipple, getRippleSizeForHorizontalComponents } from '../TouchableRipple/TouchableRipple';
 
 export type ButtonSizes = 'large' | 'medium' | 'small'
 export type ButtonTypes = 'contained' | 'outlined' | 'text'
+export type IconPositions = 'left' | 'right'
 export type TextColors = keyof IColors | 'default'
```
Comment (arielwb) [inline src/components/Button/ButtonBase.tsx:22]: Nao usar IColors!!!

File: src/components/Button/ButtonBase.tsx [hunk 7]
```diff
@@ -53,78 +54,148 @@ const getButtonStyles = (theme: Theme, type: ButtonTypes, disabled: boolean) =>
...
 const getButtonTextColor = (theme: Theme, type: ButtonTypes, disabled: boolean) => {
   const color = {
-    active: isContained(type) ? getColorOnPrimary(theme) : getColorHighEmphasis(theme),
-    disabled: isContained(type) ? getColorHighEmphasis(theme) : getColorMediumEmphasis(theme),
+    active: isContained(type) ? 'onPrimary' : 'highEmphasis',
+    disabled: isContained(type) ? 'highEmphasis' : 'mediumEmphasis',
   };
 
-  return disabled ? color.disabled : color.active;
+  return disabled ? color.disabled as keyof IColors : color.active as keyof IColors;
```
Comment (arielwb) [inline src/components/Button/ButtonBase.tsx:61]: aqui tu deveria tipar teu objeto de color, e nao forçar o tipo.
Além disso, remover IColors!!

File: src/components/TouchableRipple/TouchableRipple.tsx [hunk 8]
```diff
@@ -61,6 +62,16 @@ export const getChildrenPosition = (ref, setPosition: Function) => {
   }
 };
 
+export const getRippleSizeForHorizontalComponents = (
```
Comment (arielwb) [inline src/components/TouchableRipple/TouchableRipple.tsx:65]: O nome dessa função nao traduz o que ela faz, ela retorna a metade do tamanho maior, independente do lado.
Também acho que ela fora da lista tem responsabilidades que nao deveria ter, tipo o setSize.
Acho que ela poderia retornar o valor e deixar que quem usa decida o que fazer. Isso deixa ela mais flexivel e simplifica os testes da função.
Essa função como esta hoje não faz parte do Touchable Ripple, sugiro mover ela pra um arquivo .helpers e também seus testes.
Ah, e acho otimo ter testes para os dois casos, largura maior que altura e altura maior que largura

File: src/common/themeSelectors/sizes/sizes.ts (hunk 1 above)
Comment (lauralucca) [reply src/common/themeSelectors/sizes/sizes.ts:28]: Realmente é o caso de estarmos usando um valor que não tem token. O padding na UI DOC é 12 e tá escrito assim: 
image

Pelo que eu vejo é uma estratégia que os meninos estão usando que talvez a gente deveria explicar esse ponto que você levantou. Eu concordo que dificulta e já imagino isso criando bugs no futuro com a mudança de qualquer token.

(outra coisa que dá pra ver na imagem é esse elevation02 que nem é o nome atual do token)

**Review by arielwb — APPROVED** [review]

**Pull request merged**